oarepo_oidc_einfra/resources.py

```diff
@@ -6,50 +6,186 @@
+    def __init__(self, config: Optional[OIDCEInfraUIResourceConfig] = None):
         """Initialize the resource."""
-        super(OIDCEInfraResource, self).__init__(
-            config=config or OIDCEInfraResourceConfig()
-        )
+        super().__init__(config=config or OIDCEInfraUIResourceConfig())
 
-    def create_url_rules(self):
+    def create_url_rules(self) -> list[dict]:
+        """Create URL rules for the resource."""
+        routes = self.config.routes
+        return [
+            route("GET", routes["accept-invitation"], self.accept_invitation),
+        ]
+
+    @fresh_login_required
+    def accept_invitation(self) -> Response:
+        """Accept an invitation to join a community.
+
+        This is an endpoint to which user is directed
+        after clicking the link in the invitation email, accepting the terms and conditions and
+        accepting the invitation.
+
+        We expect the url to contain the request_id of the invitation request that was sent to the user
+        and use it to accept the invitation.
+
+        Note:
+        If user accepts the invitation but this endpoint is not called, the invitation will be forever
+        in the submitted state (until expiration). The user will still be able to access the community
+        because the AAI will return the correct capabilities for the user.
+
+        Currently, the PERUN api does not return the ID of the created invitation, so we cannot store it
+        and check in a background task if the invitation was accepted and then change the state of the request.
+
+        """
+        assert request.view_args is not None
+
+        request_id = decrypt(request.view_args["request_id"])
+
+        # force re-authentication
+        if "fresh_login_token" not in request.args:
+            # force logout and redirect to the same page with the fresh_login_token
+            # so that the user is logged in again
+            logout_user()
+            fresh_login_token = encrypt("fresh_login_token_" + str(request_id))
+            return redirect(request.url + "?fresh_login_token=" + fresh_login_token)
+        else:
+            fresh_login_token = decrypt(request.args["fresh_login_token"])
+            if fresh_login_token != "fresh_login_token_" + str(request_id):
+                raise PermissionDenied("Invalid login token")
+
+        # get the invitation request and check if it is submitted.
+        invitation_request = Request.get_record(request_id)
+        assert invitation_request.status == "submitted"
+
+        # if its user is not the current user, we might to delete
+        # the user on the request so that it does not pollute the space
+        invitation = Member.get_member_by_request(request_id)
+
+        if not invitation.model:
+            raise ValueError(f"Invitation {invitation} does not have a model.")
+
+        request_user_id = invitation.model.user_id
+
+        if str(request_user_id) != str(g.identity.id):
```

Comment:
Is it possible to also validate community roles from invitation with current user roles (coming as claims from OIDC resource caps)?

Reply:
Can not do it at this moment as the user has logged in and during this process memberships are automatically created - that is, regardless of the initial invitation, at this point the Member instance is already created and activated.

Can not do it inside the oidc as well, because at the time of login we are at the login page, not the invitation accept page.